File: components/litex_m2sdr_julia/litex_m2sdr_init.py
#!/usr/bin/env python3

import time
import argparse
import numpy as np

import SoapySDR
from SoapySDR import SOAPY_SDR_RX, SOAPY_SDR_CS16

def main():
    parser = argparse.ArgumentParser(description="LiteXM2SDR init script.")
    parser.add_argument("--samplerate", type=float, default=4e6,               help="RX Sample rate in Hz")
    parser.add_argument("--freq",       type=float, default=2.4e9,             help="RX frequency in Hz")
    parser.add_argument("--gain",       type=float, default=-20.0,             help="RX gain in dB")
    parser.add_argument("--bandwidth",  type=float, default=56e6,              help="RX filter bandwidth in Hz")
    parser.add_argument("--channel",    type=int,   choices=[0, 1], default=0, help="RX channel index (0 or 1)")
    parser.add_argument("--remote",     action="store_true",                   help="Use remote mode via SoapySDRServer (default: local)")
    parser.add_argument("--remote-ip",  type=str,   default="127.0.0.1:55132", help="Remote server IP and port (default: 127.0.0.1:55132)")

    args = parser.parse_args()

    # Prepare Remote/Local Arguments.
    if args.remote:
        device_args = {
            "driver":           "remote",
            "remote":           args.remote_ip,
            "remote:driver":    "LiteXM2SDR",
            "eth_mode":         "vrt",
        }
        mode = "remote"
    else:
        device_args = {
            "driver":   "LiteXM2SDR",
            "eth_mode": "vrt",
        }
        mode = "local"

    # Initialize the LiteXM2SDR device.
    sdr = SoapySDR.Device(','.join(f'{key}={value}' for key, value in device_args.items()))
    sdr.setSampleRate(SOAPY_SDR_RX, args.channel, args.samplerate)
    sdr.setFrequency( SOAPY_SDR_RX, args.channel, args.freq)
    sdr.setGain(      SOAPY_SDR_RX, args.channel, args.gain)
    sdr.setBandwidth( SOAPY_SDR_RX, args.channel, args.bandwidth)

    print(f"LiteXM2SDR device initialized in {mode} mode with:")
    print(f"  Sample Rate: {args.samplerate} Hz")
    print(f"  Frequency  : {args.freq} Hz")
    print(f"  Gain       : {args.gain} dB")
    print(f"  Bandwidth  : {args.bandwidth} Hz")
    print(f"  Channel    : {args.channel}")
    if args.remote:
        print(f"  Remote IP  : {args.remote_ip}\n")
    else:
        print()

    # Open and Activate the Rx Stream.
    rx_stream = sdr.setupStream(SOAPY_SDR_RX, SOAPY_SDR_CS16, [0])
    sdr.activateStream(rx_stream)

    # Buffer for receiving samples (CS16: 2x 16-bit ints = 4 bytes per sample)
    buff = np.zeros(1024, dtype=np.complex64)  # 1024 samples
    total_bytes = 0

    # Record the initialization time.
    start_time = time.time()
    last_display_time = start_time

    # Keep the device open, receive data continuously, and display every second.
    try:
        while True:
            # Read samples from the stream
            status = sdr.readStream(rx_stream, [buff], len(buff))
            if status.ret > 0:
                total_bytes += status.ret * 4  # 4 bytes per CS16 sample

            # Update display only once per second
            current_time = time.time()
            if current_time - last_display_time >= 1.0:
                elapsed = current_time - start_time
                print(f"Device has been active for {elapsed:.2f} seconds, Received: {total_bytes} bytes", end="\r")
                last_display_time = current_time

    except KeyboardInterrupt:
        print("\nExiting and closing the device.")

    # Deactivate and close the RX Stream.
    sdr.deactivateStream(rx_stream)
    sdr.closeStream(rx_stream)

if __name__ == "__main__":
    main()

File: components/litex_m2sdr_julia/litex_m2sdr_init.jl
#!/usr/bin/env -S julia --project

using ArgParse
using SoapySDR
using Unitful

# Make sure SoapySDR.jl can load system-wide plugins
ENV["SOAPY_SDR_PLUGIN_PATH"] = "/usr/lib/SoapySDR/modules0.8/"

function parse_commandline()
    s = ArgParseSettings(description="LiteXM2SDR init script.")

    @add_arg_table s begin
        "--samplerate"
            help = "RX Sample rate in Hz"
            arg_type = Float64
            default = uconvert(u"Hz", 4u"MHz").val
        "--freq"
            help = "RX frequency in Hz"
            arg_type = Float64
            default = uconvert(u"Hz", 2.4u"GHz").val
        "--gain"
            help = "RX gain in dB"
            arg_type = Float64
            default = -20
        "--bandwidth"
            help = "RX filter bandwidth in Hz"
            arg_type = Float64
            default = uconvert(u"Hz", 56u"MHz").val
        "--channel"
            help = "RX channel index (0 or 1)"
            arg_type = Int
            range_tester = x -> x in [0, 1]
            default = 0
    end

    return parse_args(s)
end

function to_frequency(val)
    freq = val*u"Hz"
    if freq >= 1e9u"Hz"
        (freq.val / 1e9) * u"GHz"
    elseif freq >= 1e6u"Hz"
        (freq.val / 1e6) * u"MHz"
    elseif freq >= 1e3u"Hz"
        (freq.val / 1e3) * u"kHz"
    else
        freq
    end
end

function main()
    args = parse_commandline()

    # Initialize the LiteXM2SDR device
    devs = Devices(driver="LiteXM2SDR")
    dev_args = devs[1]
    dev_args["eth_mode"] = "vrt"
    dev = Device(dev_args)

    # Get the RX channel
    rx_channel = dev.rx[args["channel"] + 1]

    # Apply basic configuration for the selected RX channel
    rx_channel.sample_rate = args["samplerate"] * u"Hz"
    rx_channel.frequency = args["freq"] * u"Hz"
    rx_channel.gain = args["gain"] * u"dB"
    rx_channel.bandwidth = args["bandwidth"] * u"Hz"

    println("LiteXM2SDR device initialized with:")
    println("  Sample Rate: $(to_frequency(args["samplerate"]))")
    println("  Frequency  : $(to_frequency(args["freq"]))")
    println("  Gain       : $(args["gain"]) dB")
    println("  Bandwidth  : $(to_frequency(args["bandwidth"]))")
    println("  Channel    : $(args["channel"])\n")

    # open and activate the rx stream
    rx_stream = SoapySDR.Stream(Complex{Int16}, [rx_channel])
    SoapySDR.activate!(rx_stream)

    # Keep the device open
    println("Device has been activated. Press RETURN to exit.")
    readline(stdin)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
